xapian-letor/tests/Makefile.am

```diff
@@ -19,78 +19,9 @@ libxapian_la = libxapian.la
 
 TESTS_ENVIRONMENT = ./runtest
 
-.PHONY: check-none check-inmemory \
-	check-chert check-glass \
-	check-multi check-multi-chert check-multi-glass \
-	check-remote check-remoteprog check-remotetcp \
-	check-remoteprog-chert check-remoteprog-glass \
-	check-remotetcp-chert check-remotetcp-glass \
-	up remove-cached-databases
-
-up:
-	cd .. && $(MAKE)
-
-check-none: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b none
-
-check-inmemory: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b inmemory
-
-check-multi: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b multi
-
-check-singlefile: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b singlefile
-
-if BUILD_BACKEND_REMOTE
-check-remote: check-remoteprog check-remotetcp
-
-check-remoteprog: apitest$(EXEEXT)
-## If both chert and glass are enabled, just run remoteprog_chert.
-if BUILD_BACKEND_CHERT
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remoteprog_chert
-else
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remoteprog
-endif
-
-check-remotetcp: apitest$(EXEEXT)
-## If both chert and glass are enabled, just run remotetcp_glass.
-if BUILD_BACKEND_GLASS
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remotetcp_glass
-else
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remotetcp
-endif
+.PHONY: up remove-cached-databases
 
-if BUILD_BACKEND_GLASS
-check-remoteprog-glass: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remoteprog_glass
-check-remotetcp-glass: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remotetcp_glass
-endif
-
-if BUILD_BACKEND_CHERT
-check-remoteprog-chert: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remoteprog_chert
-check-remotetcp-chert: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b remotetcp_chert
-endif
-endif
-
-if BUILD_BACKEND_GLASS
-check-multi-glass: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b multi_glass
-check-glass: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b glass
-check-singlefile-glass: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b singlefile_glass
-endif
-
-if BUILD_BACKEND_CHERT
-check-multi-chert: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b multi_chert
-check-chert: apitest$(EXEEXT)
-	$(TESTS_ENVIRONMENT) ./apitest$(EXEEXT) -b chert
-endif
+up: cd .. && $(MAKE)
```

Comment:
A phony target is one where the target name isn't a file to create, but rather effectively just a named list of commands to run.  We want `make up` should always run `cd .. && $(MAKE)`, but if there happened to be a file called `up` in the current directory `make` would see that and decide there was nothing to do.

By declaring a target as phony you tell `make` that it should always run the rule (at least for implementations such as GNU make which understand `.PHONY` - those which don't understand it will just treat `.PHONY` as another target, so you're no worse off that without the `.PHONY` list at least).

While it's pretty unlikely someone would create a file called `up`, some target names are more prone to this problem (`core` comes to mind!)

It also means `make` doesn't have to search for a match for the target in its list of implicit rules (the POSIX ones like `.c.o` and GNU make's `%.o` pattern rules), and presumably also avoids a `stat("up", &sb)` syscall to check if `up` exists and what its timestamp is - a small one-off overhead here, but might mount up for the complex tree of targets which automake creates.

I must admit I'd pretty much forgotten about `make up`, but I used to use it a lot and I likely will again now I've been reminded about it.  It's certainly the sort of thing that's good to have it supported consistently if we're going to have it.